content/issue-2/applicative-parsing.md

```diff
@@ -0,0 +1,716 @@
+}
+```
+
+If you want to write an argument parser purely imperatively, look at the horrors that entails,
+think about how to scale and maintain this:
+
+```rust
+fn parse() -> Args {
+	let bin = None;
+	let jobs = None;
+
+	let args = std::env::args();
+	while let (Some(arg), Some(value)) = (args.next(), args.next()){
+		if arg == "--bin" {
+			bin = Some(value.to_string());
+		}
+		if arg == "--jobs" {
+			jobs = Some(value.parse::<u32>.unwrap()),
+		}
+	};
+
+	Args {
+		name: name.expect("'--bin' not given"),
+		answer: answer.expect("'--jobs' not given"),
+	}
+}
+```
+
+What if you could use functional programming to just describe what you want:
+
+```rust
+fn parse() -> Args {
+    let name = Arg("--bin").parse(|n| n.to_string());
+    let answer = Arg("--jobs").parse(|a| a.parse::<u32>);
+    let Args = name
+      .zip(answer)
+      .map(|(name, answer)| Args { bin, jobs })
+      .run()
+      .unwrap()
+}
+```
+
+Now lets see how we get there.
+
+# Parsing command line options
+
+The main goal for this tutorial is to implement a basic library to parse command line options.
+If you used `cargo` you are probably familiar with expressions like this:
+
+```console
+$ cargo build --bin hello  # build a binary "hello"
+$ cargo build --jobs 4     # build everything using 4 cpu jobs
+```
+
+Parsing command line options in general is a wide problem and to make it more manageable this
+tutorial is going to look only at option arguments with a long name: `--bin hello` and
+`--jobs 4` parts. This approach however can parse pretty much anything,
+[`bpaf`](https://crates.io/crates/bpaf) is a finished library that uses it.
+
+```urlpreview
+https://crates.io/crates/bpaf
+```
+
+## Humble beginning
+
+Let's start with some magical handwavey way, taking things from `std::env::args()` and placing
+them in a container, that introduces a simple way of taking them out again.
+
+```rust
+// a magical container with a value already inside
+struct Magic(&'static str);
+impl Magic {
+    // a way to get a value out of the container
+    fn run(self) -> String {
+        self.0.into()
+    }
+}
+
+#[test]
+fn sample() {
+    // --bin hello
+    let bin = Magic("hello");
+    let result = name.run();
+    assert_eq!(result, "hello");
+}
+```
+
+At this point the parser can represent single string arguments using `Magic` and extract them via `Magic::run`.
+
+## Adding more types
+
+Next step would be to allow to have arguments of different types. Consider filenames. While
+most of the functions would take a regular string reference as a filename - not every `String` is
+a valid filename and not every filename is a valid `String`. It's a good practice to keep
+`String` for strings and `PathBuf` for filenames. Additionally, proper specialized types also serve as an anchor for documentation.
+
+To be able to consume `PathBuf` from user's input, the parser needs a way to represent it in the first
+place. This can be achieved by making `Magic` a generic datatype:
```

Comment:
consider numbers? --> jobs?

Reply:
I'm talking about numbers lower and I think I'm using `--jobs` as an example

Reply:
Is there something specific I'm still missing?

Reply:
> I'm talking about numbers lower and I think I'm using --jobs as an example

yes, you show this with numbers but you introduce it with pathbufs (and then dont use the paths).

Why not explain it with numbers (as you will expand on later with error handling) annd mention that also allows other maybe more complex types

Reply:
> Next step would be to allow to have arguments of different types. Consider `--jobs` argument.
For some applications passing "4" to the consumer is valid, but in most cases consumer would
prefer to use its numeric value. To be able to pass a number like `4u32` the parser needs a way
to represent it in the first place. This can be achieved by making `Magic` a generic datatype:

I guess... Let's see how it looks.